### ISSUE TITLE
mmkv sometimes may get wrong data

### ISSUE THREAD

**Issue by codefei** [Tencent/MMKV#361]
### The language of MMKV

Objective-C

### The version of MMKV

v1.0.19

### The platform of MMKV

iOS

### The installation of MMKV

Git clone

### What's the issue?
I use the MMKV to save the number,  but I found sometimes I get the wrong number data.

code sample like this, on the main thread
```
BOOL saveResult =[self.mmkv setInt64:0 forKey:TEST_KEY] // return value is YES, set method called success  
NSInteger value = (NSInteger)[self.mmkv getInt64ForKey:TEST_KEY] // return value is not 0(may be old data), expected is 0
```

I have read the source code, but can't find the problem. The `setRawData` method inner implementation use the `m_dic` save the data. When I use get method to read key, the return value should be equals to the value I set.

I can't reproduce the problem and can't debug the bad case. Is there anyone met the strange problem?

**Comment by lingol** [comment]
Usually, that is due to someone change the key-value somewhere else. I suggest searching for the `TEST_KEY` globally, check for each possible modification call.

**Comment by lingol** [comment]
One more thing, v1.0.19 is considerably old, you'd better upgrade to the latest version.

**Comment by codefei** [comment]
> Generally speaking, that is due to someone change the key-value somewhere else. I suggest searching for the `TEST_KEY` globally, check for each possible modification call.

I'm sure not the reason. When set value by the key, i will record it.

**Comment by codefei** [comment]
Since the value I getted is old, I doubt the `getRawDataForKey`  may get the data from file?

**Comment by lingol** [comment]
> Since the value I getted is old, I doubt the `getRawDataForKey` may get the data from file? Some situation will case `m_needLoadFromFile` is YES?

Nan, that's very likely impossible. It will take these conditions to make it happen:

1. MMKV store the new value in m_dic, and write the new value to mmap memory region.
2. Device reboot before that mmap memory is synced to file.
3. The CRC file is also not synced as well.

Then you will get the old value from MMKV after the device boots again.

**Comment by codefei** [comment]
I can reproduce it and know the reason.
Before exec the sample code above, I post the `UIApplicationDidReceiveMemoryWarningNotification` notification.
Next when I call set method, will store the value to m_dic. I first think the value is set successful.
But when call  the `appendData` method, it will check memory and will call `checkLoadData` method. 
So `loadFromFile` is executed again and loaded the old file from disk.

```
- (BOOL)setRawData:(NSData *)data forKey:(NSString *)key {
	if (data.length <= 0 || key.length <= 0) {
		return NO;
	}
	CScopedLock lock(m_lock);

	[m_dic setObject:data forKey:key];
	m_hasFullWriteBack = NO;

	return [self appendData:data forKey:key]; // it will load local old file again
}
```

**Comment by lingol** [comment]
That's remarkable. I will check & fix it soon.

**Comment by lingol** [comment]
Apparently it's already [fixed](https://github.com/Tencent/MMKV/blob/dev/iOS/MMKV/MMKV/MMKV.mm#L771) in the `dev` branch.

**Comment by lingol** [comment]
I've checked the git blame message again. In fact this bug was fixed long ago. You could just upgrade to v1.0.23 and things will be just fine.

**Comment by codefei** [comment]
> I've checked the git blame message again. In fact, this bug was fixed long ago(v1.0.20).
> You could just upgrade to the latest v1.0.23 and things will be just fine.

Thanks, Good job!
